File: console/components/shared/levels/levelTier.test.ts
import { describe, expect, it } from "vitest";
import {
  LEVEL_TIERS,
  formatTierRange,
  getLevelColor,
  getLevelTier,
  groupLevelsByTier,
  isHexColor,
} from "./levelTier";

describe("getLevelTier", () => {
  it("maps each level number to its tier of three", () => {
    expect(getLevelTier(1).name).toBe("Iron");
    expect(getLevelTier(3).name).toBe("Iron");
    expect(getLevelTier(4).name).toBe("Jade");
    expect(getLevelTier(6).name).toBe("Jade");
    expect(getLevelTier(7).name).toBe("Amethyst");
    expect(getLevelTier(9).name).toBe("Amethyst");
    expect(getLevelTier(10).name).toBe("Gold");
    expect(getLevelTier(12).name).toBe("Gold");
    expect(getLevelTier(13).name).toBe("Ruby");
    expect(getLevelTier(15).name).toBe("Ruby");
  });

  it("clamps out-of-range and invalid levels to the first/last tier", () => {
    expect(getLevelTier(0).name).toBe("Iron");
    expect(getLevelTier(-4).name).toBe("Iron");
    expect(getLevelTier(Number.NaN).name).toBe("Iron");
    expect(getLevelTier(16).name).toBe("Ruby");
    expect(getLevelTier(999).name).toBe("Ruby");
  });

  it("carries the tier hex colours", () => {
    expect(LEVEL_TIERS.map((t) => t.color)).toEqual([
      "#8A96A8",
      "#2AA876",
      "#8B5CF6",
      "#F0B429",
      "#E23D4A",
    ]);
  });
});

describe("getLevelColor", () => {
  it("prefers the level's own colour when it is a valid hex", () => {
    expect(getLevelColor({ level: 2, color: "#123456" })).toBe("#123456");
    expect(getLevelColor({ level: 2, color: " #abc " })).toBe("#abc");
  });

  it("falls back to the tier colour when colour is null, empty or invalid", () => {
    expect(getLevelColor({ level: 2, color: null })).toBe("#8A96A8");
    expect(getLevelColor({ level: 5, color: "" })).toBe("#2AA876");
    expect(getLevelColor({ level: 8, color: "purple" })).toBe("#8B5CF6");
    expect(getLevelColor({ level: 11 })).toBe("#F0B429");
    expect(getLevelColor({ level: 14, color: "#12345" })).toBe("#E23D4A");
  });
});

describe("isHexColor", () => {
  it("accepts 3 and 6 digit hex and rejects everything else", () => {
    expect(isHexColor("#fff")).toBe(true);
    expect(isHexColor("#FFAA00")).toBe(true);
    expect(isHexColor("fff")).toBe(false);
    expect(isHexColor("#ggg")).toBe(false);
    expect(isHexColor(null)).toBe(false);
    expect(isHexColor(undefined)).toBe(false);
  });
});

describe("formatTierRange", () => {
  it("formats the level range with an en dash", () => {
    expect(formatTierRange(LEVEL_TIERS[1])).toBe("Levels 4–6");
  });
});

describe("groupLevelsByTier", () => {
  it("groups in tier order, sorts within a tier and omits empty tiers", () => {
    const groups = groupLevelsByTier([
      { id: "l14", level: 14, sortOrder: 14 },
      { id: "l2", level: 2, sortOrder: 2 },
      { id: "l1", level: 1, sortOrder: 1 },
      { id: "l5", level: 5, sortOrder: 5 },
    ]);

    expect(groups.map((g) => g.tier.name)).toEqual(["Iron", "Jade", "Ruby"]);
    expect(groups[0].levels.map((l) => l.id)).toEqual(["l1", "l2"]);
    expect(groups[2].levels.map((l) => l.id)).toEqual(["l14"]);
  });

  it("orders by sortOrder before level number and returns nothing for no levels", () => {
    const groups = groupLevelsByTier([
      { id: "a", level: 1, sortOrder: 3 },
      { id: "b", level: 3, sortOrder: 1 },
    ]);
    expect(groups[0].levels.map((l) => l.id)).toEqual(["b", "a"]);
    expect(groupLevelsByTier([])).toEqual([]);
  });
});

File: mobile/test/level_tier_test.dart
import 'package:blocnet/features/levels/data/models/user_level_model.dart';
import 'package:blocnet/features/levels/domain/level_number_format.dart';
import 'package:blocnet/features/levels/domain/level_requirement.dart';
import 'package:blocnet/features/levels/domain/level_tier.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

UserLevelModel _level(int level, {String? color}) {
  return UserLevelModel(
    id: 'lvl-$level',
    slug: 'level-$level',
    name: 'Level $level',
    description: '',
    iconUrl: '',
    level: level,
    requiredBnp: '0',
    requiredComments: 0,
    requiredDaysActive: 0,
    requiredQuests: 0,
    requiredUpdates: 0,
    requiredProjects: 0,
    color: color,
    isActive: true,
    sortOrder: level,
  );
}

void main() {
  group('LevelTier.forLevel', () {
    test('maps levels 1-15 to the five tiers in order', () {
      final names = [for (var n = 1; n <= 15; n++) LevelTier.forLevel(n).name];
      expect(names, [
        'Iron', 'Iron', 'Iron',
        'Jade', 'Jade', 'Jade',
        'Amethyst', 'Amethyst', 'Amethyst',
        'Gold', 'Gold', 'Gold',
        'Ruby', 'Ruby', 'Ruby',
      ]);
    });

    test('clamps out-of-range levels instead of throwing', () {
      expect(LevelTier.forLevel(0), LevelTier.iron);
      expect(LevelTier.forLevel(-5), LevelTier.iron);
      expect(LevelTier.forLevel(16), LevelTier.ruby);
      expect(LevelTier.forLevel(99), LevelTier.ruby);
    });

    test('exposes range labels and palette', () {
      expect(LevelTier.jade.rangeLabel, 'Levels 4–6');
      expect(LevelTier.gold.fallbackColor, const Color(0xFFF0B429));
      expect(LevelTier.all.map((t) => t.index), [0, 1, 2, 3, 4]);
      expect(LevelTier.amethyst.contains(8), isTrue);
      expect(LevelTier.amethyst.contains(10), isFalse);
    });
  });

  group('tier colour resolution', () {
    test('prefers the level colour when it parses', () {
      final level = _level(5, color: '#123456');
      expect(level.tierColor, const Color(0xFF123456));
    });

    test('falls back to the tier palette when colour is missing or bad', () {
      expect(_level(5).tierColor, LevelTier.jade.fallbackColor);
      expect(_level(5, color: 'lime').tierColor, LevelTier.jade.fallbackColor);
      expect(_level(14, color: '').tierColor, LevelTier.ruby.fallbackColor);
    });

    test('tryParseHexColor accepts 6 and 8 digit forms only', () {
      expect(tryParseHexColor('#E23D4A'), const Color(0xFFE23D4A));
      expect(tryParseHexColor('E23D4A'), const Color(0xFFE23D4A));
      expect(tryParseHexColor('80E23D4A'), const Color(0x80E23D4A));
      expect(tryParseHexColor('#FFF'), isNull);
      expect(tryParseHexColor(null), isNull);
    });

    test('foregroundOn picks dark text for gold and light for the rest', () {
      expect(foregroundOn(LevelTier.gold.fallbackColor), const Color(0xFF09090B));
      expect(foregroundOn(LevelTier.ruby.fallbackColor), const Color(0xFFFFFFFF));
      expect(foregroundOn(LevelTier.iron.fallbackColor), const Color(0xFFFFFFFF));
    });
  });

  group('groupLevelsByTier', () {
    test('produces five ordered sections of three for 15 shuffled levels', () {
      final levels = [for (var n = 15; n >= 1; n--) _level(n)];
      final sections = groupLevelsByTier(levels);

      expect(sections.length, 5);
      expect(sections.map((s) => s.tier), LevelTier.all);
      for (final section in sections) {
        expect(section.levels.length, 3);
        expect(
          section.levels.map((l) => l.level),
          [section.tier.minLevel, section.tier.minLevel + 1, section.tier.maxLevel],
        );
      }
    });

    test('omits tiers with no levels and keeps order', () {
      final sections = groupLevelsByTier([_level(13), _level(2), _level(7)]);
      expect(sections.map((s) => s.tier.name), ['Iron', 'Amethyst', 'Ruby']);
    });

    test('section colour comes from the first valid level colour', () {
      final sections = groupLevelsByTier([
        _level(4, color: 'bad'),
        _level(5, color: '#2AA876'),
        _level(6, color: '#000000'),
      ]);
      expect(sections.single.color, const Color(0xFF2AA876));
    });

    test('section colour falls back to the palette', () {
      final sections = groupLevelsByTier([_level(10), _level(11)]);
      expect(sections.single.color, LevelTier.gold.fallbackColor);
    });
  });

  group('requirementsFor', () {
    test('includes only non-zero criteria and compares raw values', () {
      const level = UserLevelModel(
        id: 'l',
        slug: 'l',
        name: 'L',
        description: '',
        iconUrl: '',
        level: 9,
        requiredBnp: '1500000',
        requiredComments: 0,
        requiredDaysActive: 30,
        requiredQuests: 5,
        requiredUpdates: 2,
        requiredProjects: 1,
        color: null,
        isActive: true,
        sortOrder: 9,
      );
      const metrics = UserMetrics(
        totalBnpEarned: '1499999',
        totalComments: 999,
        totalDaysActive: 30,
        totalQuestsCompleted: 7,
        totalUpdates: 1,
        totalProjects: 0,
      );

      final rows = requirementsFor(level, metrics: metrics);
      expect(rows.map((r) => r.metric), [
        LevelMetric.bnp,
        LevelMetric.daysActive,
        LevelMetric.quests,
        LevelMetric.updates,
        LevelMetric.projects,
      ]);

      final bnp = rows.first;
      expect(bnp.isComplete, isFalse);
      expect(bnp.remaining, BigInt.one);
      expect(bnp.currentLabel, '1.4M');
      expect(bnp.requiredLabel, '1.5M');

      expect(rows[1].isComplete, isTrue); // days active 30/30
      expect(rows[2].isComplete, isTrue); // quests 7/5
      expect(rows[3].isComplete, isFalse); // updates 1/2
      expect(rows[4].remaining, BigInt.one); // projects 0/1
    });

    test('treats missing metrics as zero', () {
      final rows = requirementsFor(_level(1), metrics: null);
      expect(rows, isEmpty);
    });
  });

  group('number formatting', () {
    test('formatCompact', () {
      expect(formatCompact(BigInt.from(999)), '999');
      expect(formatCompact(BigInt.from(1000)), '1K');
      expect(formatCompact(BigInt.from(12500)), '12.5K');
      expect(formatCompact(BigInt.from(2000000)), '2M');
      expect(formatCompact(BigInt.parse('12345678901234')), '12345678.9M');
    });

    test('parseBigInt tolerates commas, decimals and junk', () {
      expect(parseBigInt('1,234'), BigInt.from(1234));
      expect(parseBigInt('12.9'), BigInt.from(12));
      expect(parseBigInt('abc'), BigInt.zero);
      expect(parseBigInt(null), BigInt.zero);
    });
  });
}
